dc/s2n-quic-dc/src/path/secret/map/state.rs

```diff
@@ -24,6 +24,119 @@ use std::{
 #[cfg(test)]
 mod tests;
 
+#[derive(Default)]
+pub(crate) struct PeerMap(
+    Mutex<hashbrown::HashTable<Arc<Entry>>>,
+    std::collections::hash_map::RandomState,
+);
+
+#[derive(Default)]
+pub(crate) struct IdMap(Mutex<hashbrown::HashTable<Arc<Entry>>>);
+
+impl PeerMap {
+    fn hash(&self, entry: &Entry) -> u64 {
+        self.hash_key(entry.peer())
+    }
+
+    fn hash_key(&self, entry: &SocketAddr) -> u64 {
+        let mut hasher = self.1.build_hasher();
+        entry.hash(&mut hasher);
+        hasher.finish()
+    }
+
+    pub(crate) fn insert(&self, entry: Arc<Entry>) -> Option<Arc<Entry>> {
+        let hash = self.hash(&entry);
+        let mut map = self.0.lock().unwrap_or_else(|e| e.into_inner());
+        match map.entry(hash, |other| other.peer() == entry.peer(), |e| self.hash(e)) {
+            hashbrown::hash_table::Entry::Occupied(mut o) => {
+                Some(std::mem::replace(o.get_mut(), entry))
+            }
+            hashbrown::hash_table::Entry::Vacant(v) => {
+                v.insert(entry);
+                None
+            }
+        }
+    }
+
+    pub(crate) fn contains_key(&self, ip: &SocketAddr) -> bool {
+        let hash = self.hash_key(ip);
+        let map = self.0.lock().unwrap_or_else(|e| e.into_inner());
+        map.find(hash, |o| o.peer() == ip).is_some()
+    }
+
+    pub(crate) fn get(&self, peer: SocketAddr) -> Option<Arc<Entry>> {
+        let hash = self.hash_key(&peer);
+        let map = self.0.lock().unwrap_or_else(|e| e.into_inner());
+        map.find(hash, |o| *o.peer() == peer).cloned()
+    }
+
+    pub(crate) fn clear(&self) {
+        let mut map = self.0.lock().unwrap_or_else(|e| e.into_inner());
+        map.clear();
+    }
+
+    pub(crate) fn retain(&self, mut filter: impl FnMut(&Arc<Entry>) -> bool) {
+        let mut map = self.0.lock().unwrap_or_else(|e| e.into_inner());
+        map.retain(|e| filter(&*e))
```

Comment:
Do we worry about the lock getting held for awhile here? Since it scans the entire table? (same for the IdMap).

Reply:
OK, I think I found a path forward. We no longer retain(...) on the ID or address maps -- in fact, the locks there are fully self-contained (no closures exposed in API). The FIFO queue lock *is* now used to ensure we've visited all entries in the map (since the FIFO queue should always contain every Entry).

That likely does impact handshake latency / s2n-quic handshake thread event loop for potentially ~5-10 milliseconds as we scan the list, but given that our dc.complete.latency is ~6ms @ p50, I think that's pretty OK -- it will likely increase the tail but not be a major effect. Scanning a Vec is also hopefully faster in the typical case -- hashbrown's iterators are not super cheap.